### PR TITLE
provide not found handler

### PR DESCRIPTION
**Pull request by xh3b4sd**
This PR provides some improvements around logging and instrumentation. This aims to help analyzing problems with microservices using microkit.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -134,18 +142,17 @@ func (s *server) Endpoints() []Endpoint {
...
 		// writing specific HTTP status codes to the given response writer can be
 		// done.
 		s.errorEncoder(ctx, err, w)
 
 		// Log the error and its errgo trace. This is really useful for debugging.
-		s.logger.Log("error", map[string]string{"domain": errDomain, "trace": errTrace, "message": errMessage})
+		errDomain := errorDomain(err)
+		errMessage := errorMessage(err)
+		errTrace := errorTrace(err)
+		s.logger.Log("error", map[string]string{"domain": errDomain, "message": errMessage, "trace": errTrace})
```
Comment (xh3b4sd) [inline server/server.go]: This is only some structural change for my inner monk.

File: server/server.go [hunk 2]
```diff
@@ -156,6 +163,7 @@ func (s *server) ErrorEncoder() kithttp.ErrorEncoder {
 		w.Header().Set("Content-Type", "application/json; charset=utf-8")
 		json.NewEncoder(w).Encode(map[string]interface{}{
 			"error": errMessage,
+			"from":  s.ServiceName(),
```
Comment (xh3b4sd) [inline server/server.go]: Errors now include information about where they are coming from.

File: server/server.go [hunk 3]
```diff
@@ -165,6 +173,36 @@ func (s *server) ErrorEncoder() kithttp.ErrorEncoder {
...
+		// This defered callback will be executed at the very end of the request.
+		defer func(t time.Time) {
+			endpointCode := strconv.Itoa(http.StatusNotFound)
+			endpointMethod := strings.ToLower(r.Method)
+			endpointName := "notfound"
+
+			endpointTotal.WithLabelValues(endpointCode, endpointMethod, endpointName).Inc()
+			endpointTime.WithLabelValues(endpointCode, endpointMethod, endpointName).Set(float64(time.Since(t) / time.Millisecond))
+
+			errorTotal.WithLabelValues(errDomain).Inc()
```
Comment (xh3b4sd) [inline server/server.go]: Not sure if we should emit error counters and normal counters together. WDYT?

File: server/server.go [hunk 4]
```diff
@@ -225,7 +263,7 @@ func (s *server) NewRouter() *mux.Router {
 			// When it is executed we know all necessary information to instrument the
 			// complete request, including its response status code.
 			defer func(t time.Time) {
-				s.logger.Log("code", endpointCode, "endpoint", endpointName, "method", endpointMethod, "path", r.URL.Path)
+				s.logger.Log("code", endpointCode, "endpoint", name, "method", endpointMethod, "path", r.URL.Path)
```
Comment (xh3b4sd) [inline server/server.go]: I would like to use the original name for logging instead the modified one which is used for metrics.

File: server/spec.go [hunk 5]
```diff
@@ -51,6 +51,9 @@ type Server interface {
 	// RequestFuncs returns the server's configured list of request functions.
 	// These are the custom request functions configured by the client.
 	RequestFuncs() []kithttp.RequestFunc
+	// ServiceName returns the name of the micro-service implementing the microkit
+	// server. This is used for logging and instrumentation.
+	ServiceName() string
```
Comment (xh3b4sd) [inline server/spec.go]: The microkit server now has to implement `ServiceName` to enrich error responses. Maybe this is also useful for other things.

File: server/server.go (hunk 3 above)
Comment (JosephSalisbury) [inline server/server.go]: This makes sense to me.

**Pull request merged**